docs/proposals/15-manage-bastions-and-ssh-key-pair-rotation.md

```diff
@@ -0,0 +1,213 @@
+- `gardenlet`
+  - Deploys `Bastion` CRD under the `extensions.gardener.cloud` API Group to the Seed, see [resource example](#resource-example) below
+  - Similar to `BackupBucket`s or `BackupEntry`, the `gardenlet` watches the `Bastion` resource in the garden cluster and creates a seed-local `Bastion` resource, on which the provider specific bastion controller acts upon
+- `gardenctlv2` (or any other client)
+  - Creates `Bastion` resource in the garden cluster
+  - Establishes an `ssh` connection to a shoot node, using a bastion host as proxy
+  - Heartbeats / keeps alive the `Bastion` resource during `ssh` connection
+- Gardener extension provider <infra>
+  - Provider specific bastion controller
+  - Should be added to gardener-extension-provider-<infra> repos, e.g. https://github.com/gardener/gardener-extension-provider-aws/tree/master/pkg/controller
+  - Has the permission to update the `Bastion/status` subresource on the seed cluster
+  - Runs on seed (of course)
+- Gardener Controller Manager (`GCM`)
+  - `Bastion` heartbeat controller
+    - Cleans up `Bastion` resource on missing heartbeat.
+    - Is configured with a `maxLifetime` and `timeToLife` for the `Bastion` resource
+- Gardener (RBAC)
+  - The project `admin` role should be extended to allow CRUD operations on the `Bastion` resource. The `gardener.cloud:system:project-member-aggregation` `ClusterRole` needs to be updated accordingly (https://github.com/gardener/gardener/blob/master/charts/gardener/controlplane/charts/application/templates/rbac-user.yaml)
+
+### SSH Flow
+0. Users should only get the RBAC permission to `create` / `update` `Bastion` resources for a namespace, if they should be allowed to `ssh` onto the shoot nodes in this namespace. A project member with `admin` role will have these permissions.
+1. User/`gardenctlv2` creates `Bastion` resource in garden cluster (see [resource example](#resource-example) below)
+    - First, gardenctl would figure out the own public IP of the user's machine. Either by calling an external service (gardenctl (v1) uses https://github.com/gardener/gardenctl/blob/master/pkg/cmd/miscellaneous.go#L226) or by calling a binary that prints the public IP(s) to stdout. The binary should be configurable. The result is set under `spec.ingress[].from[].ipBlock.cidr`
+    - the public `ssh` key of the user is set under `spec.sshPublicKey`. The key that should be used needs to be configured beforehand by the user
+    - The targeted shoot is set under `spec.shootRef`
+2. GAPI Admission Control for the `Bastion` resource in the garden cluster
+    - Mutating Webhook
+        - on creation, sets `metadata.annotations["operations.gardener.cloud/created-by"]` according to the user that created the resource
+        - when `operations.gardener.cloud/operation: keepalive` is set, calculates new `status.expirationTimestamp`
```

Comment:
> when `operations.gardener.cloud/operation: keepalive` is set

Any reason for this (and also the last heartbeat annotation) not to be a normal field in the `spec`?

Reply:
No particular reason, I can put the last heartbeat timestamp in the spec.

However the `operations.gardener.cloud/operation: keepalive` should behave similar to the `gardener.cloud/operation: foo` annotation, which is not persisted.

I will further clarify and update the documentation:
When setting `operations.gardener.cloud/operation: keepalive` it will be removed by GAPI from the annotations and updates the `spec.lastHeartbeatAt` with the current timestamp. 
The `status.expirationTimestamp` will be calculated by taking the last heartbeat timestamp and by adding `x` minutes (configurable, default 60 Minutes).

Reply:
see diff here https://github.com/gardener/gardener/pull/3802/commits/c38b2fe31bbc53d3c701790b7ce8c69c62e93bd7

Reply:
1. Just using `gardener.cloud/operation` instead of `operations.gardener.cloud/operation` is OK
2. OK, I understand, thanks. Still, I think we should have such very special behaviour as little as possible. Could the same problem also be used by simply updating the `.spec.lastHeartbeatAt` with the current timestamp (instead of setting an annotation)? Then we don't need to build this "magic annotation removal"?

Reply:
1. I have changed it to `gardener.cloud/operation`

as discussed out of band with @rfranzke we leave it as is. Otherwise we need to validate if `.spec.lastHeartbeatAt` is not too far in the future or past so that the user can't tinker with the expiration date and we let the GAPI set the correct (current) last heartbeat timestamp. (resolving conversation)